### PR TITLE
Fix #1885: Persist-first delivery for parent/child workspace messages

### LINKED ISSUE #1885: Persist-first delivery for parent/child workspace messages to prevent silent loss
## Problem

`sendMessageToParent` / `sendMessageToChild` (`src/backend/trpc/workspace.trpc.ts:569-717`) use a read-then-act pattern: find the most recent RUNNING/IDLE session, and if one exists, enqueue the message live and **skip persistence** entirely. Only when no active session is found does the flow fall back to writing a `WorkspaceNotification` row (`src/backend/orchestration/workspace-children.orchestrator.ts:86-140`).

This is a TOCTOU window: if the chosen session dies between the session lookup and `tryDispatchNextMessage`, the message is enqueued into a dead session and never persisted — silently lost, with no retry or fallback.

Notably, `fireLifecycleNotification` (`workspace-children.orchestrator.ts:147`) already avoids this by always persisting.

## Recommendation

Switch parent↔child message delivery to persist-first:

1. Always write the `WorkspaceNotification` row.
2. Attempt live delivery to an active session.
3. Mark the notification delivered on successful dispatch (the `markDelivered` / transcript-dedup machinery in `session.lifecycle.service.ts:1091-1205` already exists for exactly this).

This eliminates the message-loss window and unifies the three delivery paths (live tRPC, persistence fallback, lifecycle notifications) onto one pattern. The existing transcript-marker dedup (`<!-- factory-factory-workspace-notification:{id} -->`) should prevent double delivery when a live-delivered message is also found pending at next session start.

Context: from an architectural review of workspace/project state management.

### PR DESCRIPTION
**Pull request by martin-purplefish**
Fixes #1885.

## Problem

`sendMessageToParent` / `sendMessageToChild` used a read-then-act pattern: if an active parent/child session was found, the message was enqueued live and **never persisted**. If the chosen session died between the lookup and dispatch, the message was silently lost.

## Change

Parent↔child message delivery is now persist-first, unifying it with the lifecycle-notification path:

1. **Always persist** the `WorkspaceNotification` row first (`persistChildNotification` / `persistParentNotification` now return the created row).
2. **Attempt live delivery** to the most recent RUNNING/IDLE session, enqueuing with the `workspace-notification-{id}` message ID so the existing dispatch path (`markWorkspaceNotificationDeliveredIfNeeded`) marks the row delivered on successful send.
3. If the session dies before dispatch, the row stays pending and is delivered at the next session start; the existing transcript-marker dedup (`<!-- factory-factory-workspace-notification:{id} -->`) prevents double delivery.

Supporting refactor: the message-ID prefix, transcript marker, and enqueue-text format were previously hardcoded in three places; they now live in a shared module (`src/shared/workspace-notifications.ts`) used by the tRPC mutations, the chat dispatcher, and the session-startup redelivery path. Live delivery builds the enqueue text from the persisted row, guaranteeing it matches redelivery text exactly for content-based dedup.

Also fixed along the way: a failed live `enqueue` now leaves the notification pending and returns `delivered: false` (previously the enqueue result was ignored and the mutation reported `delivered: true`).

## Tests

- New: `workspace-children.orchestrator.test.ts` — persist functions return the created row / null on missing workspaces.
- Extended `workspace.router.test.ts` — both mutations persist before live delivery (call-order asserted), use notification-backed message IDs with the transcript marker, skip live delivery with no active session, and leave the row pending when enqueue fails.

Ran: `pnpm test` (3642 passed), `pnpm typecheck`, `pnpm check`, `pnpm check:fix`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/purplefish-ai/factory-factory/pull/1892?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

---

> [!NOTE]
> **Medium Risk**
> Changes core message-delivery ordering and adds concurrent dispatch logic for workspace notifications; behavior is heavily tested but affects agent queue/session startup paths.
> 
> **Overview**
> Parent↔child workspace messaging no longer depends on finding a live session first: **`sendMessageToParent`** and **`sendMessageToChild`** always create a **`WorkspaceNotification`** row, then optionally enqueue to the latest RUNNING/IDLE session using stable `workspace-notification-{id}` message IDs and shared enqueue text from `src/shared/workspace-notifications.ts`.
> 
> Live delivery now respects **persist-before-enqueue** ordering, returns **`delivered: false`** when enqueue fails (instead of reporting success), and skips duplicate enqueue when startup redelivery already queued the same notification. **`persistChildNotification`** / **`persistParentNotification`** return the created row or **`null`** if a workspace is missing.
> 
> The chat dispatcher gains **duplicate workspace-notification handling**: in-flight claims across sessions, drops when the row is already delivered or the message is already in the transcript, and **`removeQueuedMessage`** so phantom queue cards do not linger. Session startup redelivery uses the same shared helpers for consistent transcript-marker dedup.
> 
> Tests cover persist helpers, tRPC mutation flows, and concurrent/multi-session dispatch edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 13c63d6afd43b315774f46d1a7480490b3e8b9ff. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/shared/workspace-notifications.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+/**
+ * Build the exact user-message text enqueued for a workspace notification.
+ * Live delivery and startup redelivery must produce identical text so the
+ * transcript content match can detect an already-delivered notification.
+ */
+export function buildWorkspaceNotificationMessageText(notification: {
+  id: string;
+  direction: string;
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/shared/workspace-notifications.ts]: Fixed in 8e2fa332 — `direction` is now typed as `WorkspaceNotificationDirection` ('CHILD_TO_PARENT' | 'PARENT_TO_CHILD'), mirroring the Prisma enum, so invalid callers and future enum additions fail at compile time.

File: src/backend/trpc/workspace.trpc.ts [hunk 2]
```diff
@@ -590,56 +594,68 @@ export const workspaceRouter = router({
...
+        id: workspaceNotificationMessageId(notification.id),
+        text: buildWorkspaceNotificationMessageText(notification),
+        timestamp: new Date().toISOString(),
+        settings: {
+          selectedModel: null,
+          reasoningEffort: null,
+          thinkingEnabled: false,
+          planModeEnabled: false,
+        },
+      });
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/backend/trpc/workspace.trpc.ts:637]: Valid catch — fixed in bed1cc03. Both mutations now check `hasQueuedMessage` for the notification-backed message ID before enqueuing (same guard the startup delivery path uses), so when persist races `deliverPendingChildNotifications` on a starting session, the live path skips the duplicate enqueue and UI card and returns `delivered: true`. Covered by new tests in workspace.router.test.ts.

File: src/backend/trpc/workspace.trpc.ts [hunk 3]
```diff
@@ -590,56 +594,75 @@ export const workspaceRouter = router({
...
-      // Only persist if there is no active session to receive it live
-      if (!activeSession) {
-        await persistChildNotification({
-          parentWorkspaceId,
-          sourceWorkspaceId: input.childWorkspaceId,
-          message: input.message,
+      // Session startup delivery may have already queued this notification
+      // (persist above races deliverPendingChildNotifications on a starting session).
+      const messageId = workspaceNotificationMessageId(notification.id);
+      if (sessionDomainService.hasQueuedMessage(activeSession.id, messageId)) {
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/backend/trpc/workspace.trpc.ts:622]: Valid — the queue check alone can't catch a copy that was already dequeued. Fixed in b358c20d by adding dispatch-time dedup in `dispatchMessage` (the per-session serialization point): before sending a `workspace-notification-` message, it is dropped if (a) the same message ID is already committed to this session's transcript — airtight for same-session duplicates since the transcript commit happens inside the serialized dispatch region — or (b) the notification row is already marked delivered (covers cross-session duplicates). The check fails open so a duplicate is preferred over a lost message. Covered by new tests in chat-message-handlers.service.test.ts.

File: src/backend/services/session/service/chat/chat-message-handlers.service.ts [hunk 4]
```diff
@@ -478,6 +478,14 @@ class ChatMessageHandlerService {
     client: unknown,
     stopGeneration: number
   ): Promise<void> {
+    if (await this.isWorkspaceNotificationAlreadyDelivered(dbSessionId, msg.id)) {
+      logger.info('[Chat WS] Dropping already-delivered workspace notification', {
+        dbSessionId,
+        messageId: msg.id,
+      });
+      return;
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/backend/services/session/service/chat/chat-message-handlers.service.ts]: Valid — fixed in f43527f0. Duplicate detection moved from post-dequeue inside `dispatchMessage` to peek time in the dispatch loop: a duplicate at the head is removed via `removeQueuedMessage` (which publishes a queue snapshot) and the loop advances to the next queued message in the same pass, so nothing stalls behind a dropped duplicate. Covered by a test asserting the message behind the duplicate dispatches in the same `tryDispatchNextMessage` call.

File: src/backend/services/session/service/chat/chat-message-handlers.service.ts [hunk 5]
```diff
@@ -478,6 +478,14 @@ class ChatMessageHandlerService {
     client: unknown,
     stopGeneration: number
   ): Promise<void> {
+    if (await this.isWorkspaceNotificationAlreadyDelivered(dbSessionId, msg.id)) {
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/backend/services/session/service/chat/chat-message-handlers.service.ts]: Valid — fixed in f43527f0 (same fix as the Bugbot stall finding). The drop path now uses `removeQueuedMessage`, which emits a queue-state snapshot (no phantom card), and the dispatch loop continues to the next queued message after a drop.

File: src/backend/services/session/service/chat/chat-message-handlers.service.ts [hunk 6]
```diff
@@ -556,6 +564,46 @@ class ChatMessageHandlerService {
...
+
+    const alreadyCommitted = sessionDomainService
+      .getTranscriptSnapshot(dbSessionId)
+      .some((entry) => entry.source === 'user' && entry.id === messageId);
+    if (alreadyCommitted) {
+      return true;
+    }
+
+    try {
+      const notification = await workspaceNotificationAccessor.findById(notificationId);
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/backend/services/session/service/chat/chat-message-handlers.service.ts:666]: Valid — fixed in f43527f0. Dispatch now claims the notification in an in-process in-flight set (`claimNotificationForDispatch`) with no awaits between the duplicate checks and the claim, so the check-and-claim is atomic on the event loop. A concurrent dispatch of a duplicate copy on another session sees the claim and drops instead of double-sending; the claim is released after the send settles, so a failed send (requeued copy) can still be retried — preserving at-least-once with in-process at-most-once dispatch. All dispatch runs in the single backend process, so the set is authoritative. Covered by a test that blocks session A mid-send and asserts session B drops its copy.

File: src/backend/services/session/service/chat/chat-message-handlers.service.ts [hunk 7]
```diff
@@ -94,6 +94,13 @@ class ChatMessageHandlerService {
   /** Retry timers for provider-side busy responses that are not reflected locally yet. */
   private turnInProgressRetryTimers = new Map<string, ReturnType<typeof setTimeout>>();
   private turnInProgressRetryAttempts = new Map<string, number>();
+  /**
+   * Workspace notifications currently being delivered by some session. Claimed
+   * synchronously before dispatch so a concurrent dispatch of a duplicate copy
+   * (persist-first delivery can enqueue the same notification on two sessions)
+   * drops instead of double-sending. Released once the send settles.
+   */
+  private inFlightNotificationDeliveries = new Set<string>();
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/backend/services/session/service/chat/chat-message-handlers.service.ts]: Valid — fixed in 13c63d6a, following your suggested shape. The in-flight set is now a Map of notification ID → owning session; `resetDispatchState` (invoked by the stop handler) releases only that session's claims, so a hung send can no longer block redelivery of a still-pending notification. Release is ownership-guarded: when a stale hung dispatch finally settles after a reset, it cannot evict a claim that another session's retry now holds. Covered by two new tests — retry succeeds on another session after the claiming session is reset, and a settling stale dispatch does not release the new owner's claim.
